tests/pypi_helpers.py

```diff
@@ -0,0 +1,69 @@
+import json
```

Comment:
This file is doing way too much. It looks like it's copy a lot of logic from pip. I'd rather this code exist in another library. Maybe `pip-api` could provide this functionality. Or if not, it should probably be its own library. Is there a reason `pip download` doesn't do the trick?

You've done a lot of work here, to be sure, but I'm worried about the maintenance burden of these tests.

I was imagining some additional doctests with maybe a fixture to provide the artifacts for input.